### PR TITLE
Clean up mod1

### PR DESCRIPTION
**Pull request by eschnett**
Implement fld1 and fldmod1.
Add test cases.
Deprecate rem1.

### CODE REVIEW

File: base/deprecated.jl [hunk 1]
```diff
@@ -185,6 +185,13 @@ end
 @deprecate iround(x)              round(Integer,x)
 @deprecate iround{T}(::Type{T},x) round(T,x)
 
+# rem1 is inconsistent for x==0: The result should both have the same
```
Comment (tkelman) [inline base/deprecated.jl]: this needs to be in the "0.5 deprecations" section at the end, not mixed in with the 0.4 deprecations

File: base/operators.jl [hunk 2]
```diff
@@ -149,10 +149,11 @@ const % = rem
...
 const ÷ = div
 
-# mod returns in [0,y) whereas mod1 returns in (0,y]
+# mod returns in [0,y) or (y,0] (for negative y),
+# whereas mod1 returns in (0,y] or [y,0)
 mod1{T<:Real}(x::T, y::T) = (m=mod(x,y); ifelse(m==0, y, m))
-rem1{T<:Real}(x::T, y::T) = rem(x-1,y)+1
-fld1{T<:Real}(x::T, y::T) = fld(x-1,y)+1
+fld1{T<:Real}(x::T, y::T) = fld(x - mod1(x,y), y)
+fldmod1{T<:Real}(x::T, y::T) = (m=mod1(x,y); (fld(x-m, y), m))
```
Comment (tkelman) [inline base/operators.jl]: this should also be tested

**Comment by eschnett** [comment]
Updated.

**Comment by kshyatt** [comment]
@tkelman is this good to go?

**Comment by tkelman** [comment]
No, the docs need to be populated into the RST

**Comment by kshyatt** [comment]
Wonderful, a chance to use my new label.

**Comment by eschnett** [comment]
Updated documentation.

File: doc/stdlib/math.rst [hunk 3]
```diff
@@ -164,11 +172,29 @@ Mathematical Operators
 
    The floored quotient and modulus after division. Equivalent to ``(fld(x,y), mod(x,y))``\ .
 
+.. function:: fld1(x,m)
```
Comment (yuyichao) [inline doc/stdlib/math.rst]: Not sure if it actually makes a difference but I think we usually put space after commas in argument list.

Comment (eschnett) [inline doc/stdlib/math.rst]: So do I usually, but here I thought that this would make the tuple less readable.

Comment (yuyichao) [inline doc/stdlib/math.rst]: Just to be clear, I'm talking about the exact line I'm commenting on `.. function:: fld1(x,m)`.

Comment (eschnett) [inline doc/stdlib/math.rst]: Fixed.

File: base/docs/helpdb.jl [hunk 4]
```diff
@@ -5266,6 +5266,8 @@ doc"""
     %(x, y)
 
 Remainder from Euclidean division, returning a value of the same sign as `x`, and smaller in magnitude than `y`. This value is always exact.
+
+    x == div(x,m)*m + rem(x,m)
```
Comment (tkelman) [inline base/docs/helpdb.jl]: what's `m` ?

File: base/docs/helpdb.jl [hunk 5]
```diff
@@ -8461,13 +8465,31 @@ Matrix inverse
 inv
 
 doc"""
-    mod1(x,m)
+    fld1(x, m)
+
+Flooring division, returning a value consistent with mod1(x,m)
```
Comment (tkelman) [inline base/docs/helpdb.jl]: backticks around `mod1(x,m)`

File: doc/stdlib/math.rst (hunk 3 above)
Comment (tkelman) [inline doc/stdlib/math.rst]: signature in rst needs to be consistent

**Comment by eschnett** [comment]
Corrected documentation.

File: base/docs/helpdb.jl [hunk 6]
```diff
@@ -3265,7 +3265,7 @@ tril!(M, k)
 doc"""
     divrem(x, y)
 
-The quotient and remainder from Euclidean division. Equivalent to `(x÷y, x%y)`.
+The quotient and remainder from Euclidean division. Equivalent to `div(x,y), rem(x,y))` or `(x÷y, x%y)`.
```
Comment (tkelman) [inline base/docs/helpdb.jl]: missing open paren

File: base/docs/helpdb.jl [hunk 7]
```diff
@@ -8304,7 +8306,9 @@ procs(::SharedArray)
 doc"""
     mod(x, y)
 
-Modulus after division, returning in the range \[0,`y`), if `y` is positive, or (`y`,0\] if `y` is negative.
+Modulus after flooring division, returning in the range \[0,`y`), if `y` is positive, or (`y`,0\] if `y` is negative.
+
+    x == fld(x,m)*m + mod(x,m)
```
Comment (tkelman) [inline base/docs/helpdb.jl]: signature is in terms of y

File: doc/stdlib/math.rst (hunk 3 above)
Comment (tkelman) [inline doc/stdlib/math.rst]: rst still needs a space

File: doc/stdlib/math.rst [hunk 8]
```diff
@@ -164,11 +172,29 @@ Mathematical Operators
...
+
+   Flooring division, returning a value consistent with mod1(x,m)
+
+   .. code-block:: julia
+
+       x == fld(x,m)*m + mod(x,m)
+
+       x == (fld1(x,m)-1)*m + mod1(x,m)
+
 .. function:: mod1(x,m)
```
Comment (tkelman) [inline doc/stdlib/math.rst]: rst sig needs space for consistency

File: doc/stdlib/math.rst [hunk 9]
```diff
@@ -164,11 +172,29 @@ Mathematical Operators
...
 
    .. Docstring generated from Julia source
 
-   Modulus after division, returning in the range (0,m]
+   Modulus after flooring division, returning a value in the range (0,m]
+
+.. function:: fldmod1(x,m)
```
Comment (tkelman) [inline doc/stdlib/math.rst]: rst sig needs space for consistency

**Comment by eschnett** [comment]
I'm learning a lot about the doc system here. I don't think that's the goal here, though. Thanks for the pointers and the patience.

**Comment by tkelman** [comment]
yes, it's an open todo to make the docsystem less fiddly, the current system with partially autogenerated rst is not great.

docs lgtm now. not thrilled about adding more exports but i guess this is okay.

**Comment by eschnett** [comment]
Note that `rem1` is now deprecated, so overall the number of functions stays constant (kind of).

**Pull request merged**